File: plots/cooling_function.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib $ENV{'GALACTICUS_EXEC_PATH'         }."/perl";
use lib $ENV{'GALACTICUS_ANALYSIS_PERL_PATH'}."/perl";
use Data::Dumper;
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use GnuPlot::PrettyPlots;
use GnuPlot::LaTeX;

# Make a plot of the specified transfer function file.
# Andrew Benson (27-Jan-2009)

die "Usage: cooling_function.pl <coolingFunctionFile> <pdfFile>"
    unless ( scalar(@ARGV) == 1 || scalar(@ARGV) == 2 );
my $coolingFunctionFile = $ARGV[0];
my $pdfFile             = $ARGV[1];

# Read the data file.
my $data = new PDL::IO::HDF5($coolingFunctionFile);
(my $description)    = $data->attrGet('description')       ;
my $coolingFunctions = $data->dataset('coolingRate')->get();
my $metallicities    = $data->dataset('metallicity')->get();
my $temperatures     = $data->dataset('temperature')->get();

# Create plot titles.
my $lZMin = +100.0;
my $lZMax = -100.0;
my @plotTitles;
for (my $iMetallicity=0;$iMetallicity<nelem($metallicities);++$iMetallicity) {
    my $metallicity = $metallicities->(($iMetallicity));
    if ( $metallicity <= -999.0 ) {
	push(@plotTitles,"Primordial");
    } else {
	push(@plotTitles,"\$\\log_{10}(Z/Z_\\odot) = ".$metallicity."\$");
	if ( $metallicity > $lZMax ) {$lZMax=$metallicity};
	if ( $metallicity < $lZMin ) {$lZMin=$metallicity};
    }
}

# Make a plot of the cooling function.
my $plot;
my $gnuPlot;
my $plotFile = $pdfFile;
(my $plotFileTeX = $plotFile) =~ s/\.pdf$/.tex/;
open($gnuPlot,"|gnuplot 1>/dev/null 2>&1");
print $gnuPlot "set terminal cairolatex pdf standalone color lw 2 size 5.0in,3.3in\n";
print $gnuPlot "set output '".$plotFileTeX."'\n";
print $gnuPlot "set title offset 0,-1 '".$description.", colored by \$\\log_{10}(Z/Z_\\odot)\$'\n";
print $gnuPlot "set xlabel 'Temperature [K]'\n";
print $gnuPlot "set ylabel '\$\\Lambda(T)\$ [erg cm\$^3\$ s\$^{-1}\$]'\n";
print $gnuPlot "set lmargin screen 0.15\n";
print $gnuPlot "set rmargin screen 0.85\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key off\n";
print $gnuPlot "set logscale xy\n";
print $gnuPlot "set mxtics 10\n";
print $gnuPlot "set mytics 10\n";
print $gnuPlot "set format x '\$10^{\%L}\$'\n";
print $gnuPlot "set format y '\$10^{\%L}\$'\n";
print $gnuPlot "set palette rgbformulae 33,13,10\n";
print $gnuPlot "set cbrange [".$lZMin.":".$lZMax."]\n";
print $gnuPlot "set xrange [316.0:1.0e9]\n";
print $gnuPlot "set yrange [3.0e-30:1.0e-19]\n";
print $gnuPlot "set pointsize 2.0\n";
# Loop over cooling functions.
for (my $iMetallicity=0;$iMetallicity<nelem($metallicities);++$iMetallicity) {
    my $cFrac = $iMetallicity/(nelem($metallicities)-1);
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot,
	$temperatures,
	$coolingFunctions->(($iMetallicity),:),
	style  => "line",
	weight => [3,1],
	color  => ["palette frac ".$cFrac,"palette frac ".$cFrac]
	);
}
&GnuPlot::PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&GnuPlot::LaTeX::GnuPlot2PDF($plotFileTeX);

exit;

File: plots/cooling_function.py
#!/usr/bin/env python3
import sys
import numpy as np
import h5py
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

# Make a plot of the specified cooling function file.
# Andrew Benson (27-Jan-2009) [Perl]; ported to Python

if len(sys.argv) < 2 or len(sys.argv) > 3:
    print("Usage: cooling_function.py <coolingFunctionFile> [<pdfFile>]", file=sys.stderr)
    sys.exit(1)

coolingFunctionFile = sys.argv[1]
pdfFile             = sys.argv[2] if len(sys.argv) == 3 else "cooling_function.pdf"

# Read the data file.
with h5py.File(coolingFunctionFile, 'r') as data:
    description     = data.attrs['description']
    if isinstance(description, bytes):
        description = description.decode()
    coolingFunctions = data['coolingRate'][()]
    metallicities    = data['metallicity'][()]
    temperatures     = data['temperature'][()]

# Build plot titles and find metallicity range (excluding primordial).
lZMin = +1.0e100
lZMax = -1.0e100
for iMetallicity, metallicity in enumerate(metallicities):
    if metallicity > -999.0:
        if metallicity > lZMax:
            lZMax = metallicity
        if metallicity < lZMin:
            lZMin = metallicity
useColorBar = lZMin < lZMax
            
# Make the plot.
fig, ax = plt.subplots(figsize=(5.0, 3.3))

nMetallicities = len(metallicities)
if useColorBar:
    cmap = plt.get_cmap('rainbow')
    norm = mcolors.Normalize(vmin=lZMin, vmax=lZMax)

for iMetallicity in range(nMetallicities):
    metallicity = metallicities[iMetallicity]
    if useColorBar:
        if metallicity <= -999.0:
            color = cmap(0.0)
        else:
            color = cmap(norm(metallicity))
    else:
        color = 'red'
    ax.plot(temperatures, coolingFunctions[:, iMetallicity], color=color, linewidth=1.5)

ax.set_xscale('log')
ax.set_yscale('log')
ax.set_xlim(316.0, 1.0e9)
ax.set_ylim(3.0e-30, 1.0e-19)
ax.set_xlabel('Temperature [K]')
ax.set_ylabel(r'$\Lambda(T)$ [erg cm$^3$ s$^{-1}$]')
ax.set_title(description)

if useColorBar:
    sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)
    sm.set_array([])
    fig.colorbar(sm, ax=ax, label=r'$\log_{10}(Z/Z_\odot)$')

plt.tight_layout()
plt.savefig(pdfFile)
plt.close()
